RFC-0032-numpy-support-in-dynamo.md

```diff
@@ -0,0 +1,380 @@
+implement support for their arguments in a generic way as a preprocessing or
+postprocessing step.
+
+**The `ndarray` class** Once we have all the free functions implemented as
+functions from `torch_np.ndarray`s to `torch_np.ndarray`s, implementing the
+methods from the `ndarray` class is rather simple. We simply register all the
+free functions as methods or dunder methods appropriately. We also forward the
+properties of `ndarray to the corresponding properties of `torch.Tensor` and we
+are done. This creates a circular dependency which we break with a local
+import.
+
+### Testing
+
+The testing of the framework was done via ~~copying~~ vendoring tests from the
+NumPy test suite. Then, we would replace the NumPy imports with `torch_np`
+imports. The failures on these tests were then triaged, and either fixed or marked
+`xfail` depending on our assessment of the priority of implementing a fix.
+
+In the end, to have a last check that this tool was sound, we pulled five
+examples of NumPy code from different sources and ran it with this library (eager mode execution).
+We were able to run the five examples successfully with close to no code changes.
+You can read about these in the [README](https://github.com/Quansight-Labs/numpy_pytorch_interop).
+
+### Limitations
+
+A number of known limitations are tracked in the second part of the
+[OP of this issue](https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/73).
+When landing this RFC, we will create a comprehensive document with the differences
+between NumPy and `torch_np`.
+
+### Beyond plain NumPy
+
+**GPU**. The current implementation so far only been implemented and tested on
+CPU. We expect GPU coverage to be as good as the coverage we have with CPU-GPU
+matching tests in the PyTorch test suite. If the original tensors are on GPU,
+the execution should be performed fully on GPU.
+
+**Gradients**. We have not tested gradient tracking either as we are still to
+find some good examples on which to test it, but it should be a simple
+corollary of all this effort. If the original tensors fed into a function
+have `requires_grad=True`, the tensors will track the gradients of the internal
+implementation and then the user can differentiate through their NumPy code.
+
+### Bindings to TorchDynamo
+
+The bindings for NumPy at the TorchDynamo level are currently being developed in
+[pytorch#95849](https://github.com/pytorch/pytorch/pull/95849).
+
+
+## Unresolved questions
+
+A question was left open in the initial discussion. Should the module
+`torch_np` be publicly exposed as `torch.numpy` or not?
+
+A few arguments in favor of making it public:
+* People could use it in their NumPy programs just by changing the import to
+  `import torch.numpy as np`. This could be a selling point similar to JAX's
+  `jax.numpy`, which could incentivize adoption.
+* People would not need to use the whole PyTorch 2.0 stack to start using
```

Comment:
Not sure what this sentence means?

Reply:
The idea here is that using `torch.compile` already adds a fair amount of technical overhead: when something doesn't work, you get huge errors, sometimes informative, sometimes not. If this API were public, you could first test it just by changing the relevant imports, and if anything breaks, you'll get a readable error. Then, once you know your implementation is sound in eager mode, you can simply add the `torch.compile` decorator and have it sped up

Reply:
Ho I misread the "not" here I think :D
Maybe replace the end to "to start using PyTorch as a backend for their numpy code in their codebases".

Reply:
Also we should update `PyTorch 2.0` -> `torch.compile` :D

Reply:
done and done